library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java

```diff
@@ -121,136 +120,109 @@ private void createSetters()
-      }
+    List<Object> key = new ArrayList<>();
+    String[] reqKeys = index == 0 ? leftKeys : rightKeys;
+    Map<String,PojoUtils.Getter> gettersStream = index == 0 ? gettersStream1 : gettersStream2;
+    for (String lkey : reqKeys ) {
+      key.add(gettersStream.get(lkey).get(input));
     }
-    return map;
+    return key;
   }
 
   @Override
-  public List<List<Map<String, Object>>> merge(List<List<Map<String, Object>>> accumulatedValue1, List<List<Map<String, Object>>> accumulatedValue2)
+  public List<Multimap<List<Object>, Object>> merge(List<Multimap<List<Object>, Object>> accumulatedValue1, List<Multimap<List<Object>, Object>> accumulatedValue2)
   {
     for (int i = 0; i < 2; i++) {
-      List<Map<String, Object>> curList = accumulatedValue1.get(i);
-      curList.addAll(accumulatedValue2.get(i));
-      accumulatedValue1.set(i, curList);
+      Multimap<List<Object>, Object> mMap1 = accumulatedValue1.get(i);
+      Multimap<List<Object>, Object> mMap2 = accumulatedValue2.get(i);
+      for (Map.Entry<List<Object>, Object> entry : mMap2.entries()) {
+        mMap1.put(entry.getKey(),entry.getValue());
+      }
     }
     return accumulatedValue1;
   }
 
   @Override
-  public List<?> getOutput(List<List<Map<String, Object>>> accumulatedValue)
+  public List<?> getOutput(List<Multimap<List<Object>, Object>> accumulatedValue)
   {
     if (setters == null) {
       createSetters();
-      keySetStream2 = new HashSet<>();
-      keySetStream1 = new HashSet<>();
-      int lIndex = getLeftStreamIndex();
-      for (int i = 0; i < leftKeys.length; i++) {
-        keySetStream1.add(lIndex == 0 ? leftKeys[i] : rightKeys[i]);
-        keySetStream2.add(lIndex == 1 ? leftKeys[i] : rightKeys[i]);
-      }
     }
-
     // TODO: May need to revisit (use state manager).
-    List<Map<String, Object>> result = getAllCombo(0, accumulatedValue, null);
+    return getAllCombo(accumulatedValue);
+  }
 
-    List<Object> out = new ArrayList<>();
-    for (Map<String, Object> resultMap : result) {
-      Object o;
-      try {
-        o = outClass.newInstance();
-      } catch (Throwable e) {
-        throw Throwables.propagate(e);
-      }
-      for (Map.Entry<String, PojoUtils.Setter> setter : setters.entrySet()) {
-        if (resultMap.get(setter.getKey()) != null) {
-          setter.getValue().set(o, resultMap.get(setter.getKey()));
-        }
+  public void setObjectForResult(Map<String,PojoUtils.Getter> stream, Object input, Object output)
+  {
+    for (Map.Entry<String, PojoUtils.Getter> getter : stream.entrySet()) {
+      if (setters.containsKey(getter.getKey())) {
+        setters.get(getter.getKey()).set(output, getter.getValue().get(input));
       }
-      out.add(o);
     }
 
-    return out;
   }
 
-
-  public List<Map<String, Object>> getAllCombo(int streamIndex, List<List<Map<String, Object>>> accu, Map<String, Object> curMap)
+  public List<Object> getAllCombo(List<Multimap<List<Object>, Object>> accu)
```

Comment:
For the sake of readers of this accumulation, can you describe the logic in this method and how it handles all case of joins?